### PR TITLE
refactor: updated attestation logic to the new algorithm

### PR DESCRIPTION
**Pull request by pompon0**
The new algorithm  (https://github.com/matter-labs/era-consensus/pull/175)
supports dynamic attestation committee. This pr does NOT implement committee rotation,
it will be done once the consensus registry contract is merged.

ENs no longer store the certificates.
I've also implemented a test of the attestation logic.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -300,3 +300,17 @@ zksync_contract_verification_server = { version = "0.1.0", path = "core/node/con
...
+zksync_consensus_bft = { path = '/home/gprusak/github/era-consensus/node/actors/bft' }
+zksync_consensus_crypto = { path = '/home/gprusak/github/era-consensus/node/libs/crypto' }
+zksync_consensus_executor = { path = '/home/gprusak/github/era-consensus/node/actors/executor' }
+zksync_consensus_network = { path = '/home/gprusak/github/era-consensus/node/actors/network' }
+zksync_consensus_roles = { path = '/home/gprusak/github/era-consensus/node/libs/roles' }
+zksync_consensus_storage = { path = '/home/gprusak/github/era-consensus/node/libs/storage' }
+zksync_consensus_utils = { path = '/home/gprusak/github/era-consensus/node/libs/utils' }
+zksync_concurrency = { path = '/home/gprusak/github/era-consensus/node/libs/concurrency' }
+zksync_protobuf = { path = '/home/gprusak/github/era-consensus/node/libs/protobuf' }
+zksync_protobuf_build = { path = '/home/gprusak/github/era-consensus/node/libs/protobuf_build' }
```
[the remark this answers is not in the document]
Comment (pompon0) [reply Cargo.toml:314]: of course, I'll fix that once the other pr is merged.

[the remark this answers is not in the document]
Comment (aakoshh) [reply Cargo.toml:314]: Ah ok; I pointed it at the last commit of the open PR, but it's close to being merged anyway.

File: core/node/consensus/src/en.rs [hunk 2]
```diff
@@ -174,6 +162,62 @@ impl EN {
...
+                    batch_to_attest: attester::Batch {
+                        genesis: status.genesis,
+                        hash,
+                        number: status.next_batch_to_attest,
+                    },
+                    committee: committee.clone(),
+                }))
+                .await
+                .context("start_attestation()")?;
+            next = status.next_batch_to_attest.next();
```
Comment (aakoshh) [inline core/node/consensus/src/en.rs:217]: I can't see the EN saving the certificate any more. Is that deliberate?

File: core/node/consensus/src/storage/connection.rs [hunk 3]
```diff
@@ -47,6 +47,28 @@ impl ConnectionPool {
             ctx.sleep(POLL_INTERVAL).await?;
         }
     }
+
+    /// Waits for the `number` L1 batch hash.
+    pub async fn wait_for_batch_hash(
```
Comment (aakoshh) [inline core/node/consensus/src/storage/connection.rs:52]: I reckon this is a method that many would like to see instrumented, to learn how long it takes to wait for the hash.

**Review by aakoshh — COMMENTED** [review]
Looks good! 

Can't say I'll miss those methods that were removed from the storages 👏

File: core/node/consensus/src/en.rs (hunk 2 above)
Comment (pompon0) [reply core/node/consensus/src/en.rs:217]: yes

Comment (aakoshh) [reply core/node/consensus/src/en.rs:217]: Okay, might want to mention it in the PR description. @RomanBrodetski so from now you can only look for certificates on the main node.

File: core/node/consensus/src/storage/connection.rs (hunk 3 above)
Comment (pompon0) [reply core/node/consensus/src/storage/connection.rs:52]: done

File: core/node/consensus/src/en.rs (hunk 2 above)
Comment (pompon0) [reply core/node/consensus/src/en.rs:217]: added to description

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (pompon0) [reply Cargo.toml]: done

**Review by RomanBrodetski — APPROVED** [review]

**Pull request merged**